test/Tests/InferDataContractResolverTests.cs

```diff
@@ -0,0 +1,236 @@
+            }
+
+            foreach (var field in type.GetFields(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.DeclaredOnly))
+            {
+                VisitTypesRecursively(field.FieldType, visitType, alreadyReviewed);
+            }
+        }
+
+        [Fact]
+        public void InferDataContractResolverTests_RecoversTypes()
+        {
+            var alreadyTested = new HashSet<Type>();
+
+            TestCase(typeof(double[,]));
+
+            TestCase(typeof(Tuple<Tuple<List<string>, Tuple<int>>>));
+            TestCase(typeof(string));
+            TestCase(typeof(string[]));
+            TestCase(typeof(List<string>));
+            TestCase(typeof(X<string>.Y<int[]>.Z<char>));
+            TestCase(typeof(X<Tuple<List<string>, Tuple<int[]>>>.Y<Tuple<List<string>, Tuple<int>>>.Z<Tuple<List<string>, Tuple<int>[]>>));
+
+            TestAssembly(typeof(string).Assembly);
+            TestAssembly(typeof(Gaussian).Assembly);
+
+            void TestAssembly(Assembly assembly)
+            {
+                var types = assembly.GetTypes();
+                VisitTypesRecursively(types, TestCase);
+            }
+
+            void TestCase(Type type)
+            {
+                // Avoid repeating tests.
+                if (!alreadyTested.Add(type))
+                {
+                    return;
+                }
+
+                var mentionedTypes = InferDataContractResolver.GetTypeLayout(type.FullName);
+
+                var typeTree = AddMentionedTypes(type);
+
+                AssertAreEqual(mentionedTypes, typeTree);
+
+                void AssertAreEqual(InferDataContractResolver.TypeId x, InferDataContractResolver.TypeId y)
+                {
+                    if (x.Name != y.Name)
+                    {
+                        throw new InvalidOperationException("Incorrect name");
+                    }
+
+                    if (x.ArrayLayout.Length != y.ArrayLayout.Length)
+                    {
+                        throw new InvalidOperationException($"Incorrect isArray");
+                    }
+
+                    for (int i = 0; i < x.ArrayLayout.Length; i++)
+                    {
+                        if (x.ArrayLayout[i] != y.ArrayLayout[i])
+                        {
+                            throw new InvalidOperationException($"Incorrect array layout entry");
+                        }
+                    }
+
+                    if (x.Arguments.Length != y.Arguments.Length)
+                    {
+                        throw new InvalidOperationException($"Incorrect length");
+                    }
+
+                    for (int i = 0; i < x.Arguments.Length; i++)
+                    {
+                        AssertAreEqual(x.Arguments[i], y.Arguments[i]);
+                    }
+                }
+
+                InferDataContractResolver.TypeId AddMentionedTypes(Type mentionedType)
```

Comment:
Should be called `GetTypeId`

Reply:
Changed to "GetParsedTypeString" to be consistent with rename of TypeId to ParsedTypeString.